File: sdks/android/mm-sdk/src/main/kotlin/com/matrixmedia/sdk/internal/auth/TokenManager.kt
package com.matrixmedia.sdk.internal.auth

import com.matrixmedia.sdk.MMAuthToken
import com.matrixmedia.sdk.MMException
import com.matrixmedia.sdk.MMOpenIDToken
import com.matrixmedia.sdk.MMRefreshResponse
import com.matrixmedia.sdk.MMTokenResponse
import com.matrixmedia.sdk.MMUser
import com.matrixmedia.sdk.internal.networking.MMApiClient
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Manages MM session token lifecycle: authentication, storage, and automatic refresh.
 *
 * ## Token Strategy (PTT lesson: never cache stale tokens)
 *
 * The [tokenProvider] is the external token source (Matrix OpenID) and is called
 * during initial authentication. The MM session token (JWT with 15-minute TTL)
 * is managed internally with auto-refresh at 80% TTL (i.e., refresh at 12 minutes).
 *
 * The refresh token (24-hour TTL) is used to obtain a new session token without
 * requiring a fresh Matrix OpenID exchange.
 *
 * ## Concurrency
 *
 * A [Mutex] guards token refresh to prevent thundering-herd on concurrent API calls.
 * Multiple callers to [getToken] will all await the same refresh operation.
 *
 * @param serverUrl MM server base URL.
 * @param tokenProvider External suspend function providing Matrix OpenID tokens.
 */
internal class TokenManager(
    private val serverUrl: String,
    private val tokenProvider: suspend () -> MMAuthToken
) {
    private val mutex = Mutex()

    private var sessionToken: String? = null
    private var refreshToken: String? = null
    private var expiresAtMs: Long = 0L
    private var authenticatedUser: MMUser? = null

    // Lazy initialization -- set by MMClient after construction
    internal var apiClient: MMApiClient? = null

    private val moshi: Moshi = Moshi.Builder()
        .addLast(KotlinJsonAdapterFactory())
        .build()

    /**
     * Authenticate with the MM server by exchanging a Matrix OpenID token.
     *
     * POST /_mm/client/v1/auth/token with the OpenID token. The server validates
     * against the homeserver and returns an MM session JWT + refresh token.
     *
     * @param openIdToken Matrix OpenID token from the homeserver.
     * @return The authenticated [MMUser].
     */
    suspend fun authenticate(openIdToken: MMOpenIDToken): MMUser = mutex.withLock {
        val client = requireApiClient()

        val body = buildString {
            append("{")
            append("\"access_token\":\"${openIdToken.accessToken}\"")
            append(",\"token_type\":\"${openIdToken.tokenType}\"")
            append(",\"matrix_server_name\":\"${openIdToken.matrixServerName}\"")
            append(",\"expires_in\":${openIdToken.expiresIn}")
            append("}")
        }

        val responseBody = client.exchangeToken(body)
        val adapter = moshi.adapter(MMTokenResponse::class.java)
        val response = adapter.fromJson(responseBody)
            ?: throw MMException.Server("MM_PARSE_ERROR", "Failed to parse auth response")

        sessionToken = response.accessToken
        refreshToken = response.refreshToken
        expiresAtMs = System.currentTimeMillis() + (response.expiresInSeconds * 1000L)

        val user = MMUser(
            userId = response.userId,
            displayName = response.displayName,
            avatarUrl = response.avatarUrl
        )
        authenticatedUser = user
        user
    }

    /**
     * Get a valid session token, refreshing if necessary.
     *
     * Refreshes automatically when the token is at 80% of its TTL or expired.
     * Uses a mutex to prevent concurrent refresh storms.
     *
     * @return A valid MM session JWT.
     * @throws MMException.NotAuthenticated if [authenticate] has not been called.
     * @throws MMException.TokenRefreshFailed if refresh fails.
     */
    suspend fun getToken(): String = mutex.withLock {
        val token = sessionToken ?: throw MMException.NotAuthenticated()

        // Refresh at 80% TTL to avoid using tokens close to expiry
        val refreshThresholdMs = expiresAtMs - ((expiresAtMs - (expiresAtMs - 900_000L)) * 0.2).toLong()
        val now = System.currentTimeMillis()

        if (now < refreshThresholdMs) {
            return token
        }

        // Token needs refresh
        return refreshSessionToken()
    }

    /**
     * Ensure the client is authenticated. Throws if not.
     */
    fun ensureAuthenticated() {
        if (sessionToken == null) {
            throw MMException.NotAuthenticated()
        }
    }

    /**
     * Clear the session, invalidating all tokens.
     */
    fun clearSession() {
        sessionToken = null
        refreshToken = null
        expiresAtMs = 0L
        authenticatedUser = null
    }

    /**
     * Refresh the session token using the refresh token.
     */
    private suspend fun refreshSessionToken(): String {
        val refresh = refreshToken
            ?: throw MMException.TokenRefreshFailed(IllegalStateException("No refresh token"))

        val client = requireApiClient()

        return try {
            val responseBody = client.refreshToken(refresh)
            val adapter = moshi.adapter(MMRefreshResponse::class.java)
            val response = adapter.fromJson(responseBody)
                ?: throw MMException.TokenRefreshFailed(IllegalStateException("Failed to parse refresh response"))

            sessionToken = response.accessToken
            refreshToken = response.refreshToken
            expiresAtMs = System.currentTimeMillis() + (response.expiresInSeconds * 1000L)

            response.accessToken
        } catch (e: MMException) {
            throw e
        } catch (e: Exception) {
            // Refresh failed -- clear session so the caller can re-authenticate
            clearSession()
            throw MMException.TokenRefreshFailed(e)
        }
    }

    private fun requireApiClient(): MMApiClient {
        return apiClient ?: throw IllegalStateException(
            "TokenManager.apiClient not set. This is an SDK internal error."
        )
    }
}

File: sdks/ios/Sources/MatrixMediaSDK/Internal/Auth/TokenManager.swift
import Foundation

/// Manages the MM session JWT lifecycle: authentication, caching, and auto-refresh.
///
/// ## Design (PTT Knowledge Base lesson #9)
///
/// The TokenManager holds two types of tokens:
/// 1. **Matrix OpenID token** -- obtained via the `tokenProvider` closure. NEVER cached;
///    the closure is called fresh each time authentication is needed.
/// 2. **MM session JWT** -- obtained by exchanging the OpenID token with mm-core.
///    Cached internally with auto-refresh at 80% of TTL.
///
/// The `getToken()` method is the single entry point for all authenticated API calls.
/// It returns a valid session JWT, refreshing transparently if needed.
final class TokenManager: @unchecked Sendable {

    // MARK: - Constants

    /// Refresh the token when 80% of its TTL has elapsed.
    /// For a 15-minute JWT, refresh happens at 12 minutes.
    private static let refreshThreshold: Double = 0.80

    // MARK: - Properties

    private let serverURL: URL
    private let tokenProvider: @Sendable () async throws -> MMOpenIDToken
    private let lock = NSLock()

    /// Current session state, protected by `lock`.
    private var currentToken: MMAuthToken?
    private var userId: String?
    private var displayName: String?

    /// Prevents concurrent refresh attempts.
    private var refreshTask: Task<MMAuthToken, Error>?

    // MARK: - Lazy API Client

    /// Lazy reference to avoid circular dependency (MMAPIClient holds TokenManager).
    /// Set by MMAPIClient after construction.
    private var _apiClient: MMAPIClient?

    var apiClient: MMAPIClient {
        get {
            guard let client = _apiClient else {
                fatalError("TokenManager.apiClient not set. This is a programming error.")
            }
            return client
        }
        set { _apiClient = newValue }
    }

    // MARK: - Init

    init(serverURL: URL, tokenProvider: @escaping @Sendable () async throws -> MMOpenIDToken) {
        self.serverURL = serverURL
        self.tokenProvider = tokenProvider
    }

    // MARK: - Public API

    /// Authenticate with mm-core by obtaining a fresh OpenID token and exchanging it.
    ///
    /// - Returns: The authenticated ``MMUser``.
    func authenticate() async throws -> MMUser {
        // 1. Get a fresh OpenID token (never cached)
        let openIDToken: MMOpenIDToken
        do {
            openIDToken = try await tokenProvider()
        } catch {
            throw MMError.notAuthenticated
        }

        // 2. Exchange for MM session JWT
        let authToken = try await apiClient.exchangeToken(openIDToken: openIDToken)

        // 3. Store session
        lock.lock()
        currentToken = authToken
        userId = authToken.userId
        lock.unlock()

        return MMUser(
            id: authToken.userId,
            displayName: nil,
            avatarURL: nil,
            homeserver: openIDToken.matrixServerName
        )
    }

    /// Get a valid session JWT for API requests.
    ///
    /// If the current token is expired or about to expire, transparently refreshes it.
    /// If refresh fails, re-authenticates from scratch via the token provider.
    ///
    /// - Returns: A valid MM session JWT string.
    func getToken() async throws -> String {
        lock.lock()
        guard let token = currentToken else {
            lock.unlock()
            throw MMError.notAuthenticated
        }

        // Check if token needs refresh (80% of TTL elapsed)
        let now = Date()
        let ttl = token.expiresAt.timeIntervalSince(now)
        let totalTTL = token.expiresAt.timeIntervalSince(
            token.expiresAt.addingTimeInterval(-900) // Assume 15-min TTL for threshold
        )
        let threshold = totalTTL * Self.refreshThreshold

        if ttl > (totalTTL - threshold) && ttl > 0 {
            // Token is still valid and not near expiry
            let jwt = token.sessionToken
            lock.unlock()
            return jwt
        }

        // Token needs refresh
        if let existing = refreshTask {
            lock.unlock()
            let refreshed = try await existing.value
            return refreshed.sessionToken
        }

        let refreshTokenValue = token.refreshToken
        let task = Task<MMAuthToken, Error> {
            defer {
                self.lock.lock()
                self.refreshTask = nil
                self.lock.unlock()
            }

            do {
                // Try refresh first
                let newToken = try await self.apiClient.refreshToken(refreshToken: refreshTokenValue)
                let updatedToken = MMAuthToken(
                    sessionToken: newToken.sessionToken,
                    refreshToken: newToken.refreshToken,
                    expiresAt: newToken.expiresAt,
                    userId: self.userId ?? ""
                )
                self.lock.lock()
                self.currentToken = updatedToken
                self.lock.unlock()
                return updatedToken
            } catch {
                // Refresh failed -- try full re-authentication
                do {
                    let openIDToken = try await self.tokenProvider()
                    let authToken = try await self.apiClient.exchangeToken(openIDToken: openIDToken)
                    self.lock.lock()
                    self.currentToken = authToken
                    self.userId = authToken.userId
                    self.lock.unlock()
                    return authToken
                } catch {
                    self.lock.lock()
                    self.currentToken = nil
                    self.lock.unlock()
                    throw MMError.tokenRefreshFailed(underlying: error)
                }
            }
        }

        refreshTask = task
        lock.unlock()
        let result = try await task.value
        return result.sessionToken
    }

    /// Clear all stored session state.
    func clearSession() {
        lock.lock()
        currentToken = nil
        userId = nil
        displayName = nil
        refreshTask?.cancel()
        refreshTask = nil
        lock.unlock()
    }

    /// Whether the manager currently holds a session token (may be expired).
    var hasSession: Bool {
        lock.lock()
        defer { lock.unlock() }
        return currentToken != nil
    }
}
